source/locationHelper.py

```diff
@@ -0,0 +1,437 @@
+
+	def toRECT(self):
+		"""Converts self to a L{ctypes.wintypes.RECT}"""
+		return RECT(self.left,self.top,self.right,self.bottom)
+
+	def toLogical(self, hwnd):
+		left,top=self.topLeft.toLogical(hwnd)
+		right,bottom=self.bottomRight.toLogical(hwnd)
+		if isinstance(self, RectLTWH):
+			return RectLTWH(left,top,right-left,bottom-top)
+		return RectLTRB(left,top,right,bottom)
+
+	def toPhysical(self, hwnd):
+		left,top=self.topLeft.toPhysical(hwnd)
+		right,bottom=self.bottomRight.toPhysical(hwnd)
+		if isinstance(self, RectLTWH):
+			return RectLTWH(left,top,right-left,bottom-top)
+		return RectLTRB(left,top,right,bottom)
+
+	def toClient(self, hwnd):
+		left, top =self.topLeft.toClient(hwnd)
+		if isinstance(self, RectLTWH):
+			return RectLTWH(left, top, self.width, self.height)
+		return RectLTRB(left, top, left+self.width, top+self.height)
+
+	def toScreen(self, hwnd):
+		left,top=self.topLeft.toScreen(hwnd)
+		if isinstance(self, RectLTWH):
+			return RectLTWH(left, top, self.width, self.height)
+		return RectLTRB(left, top, left+self.width, top+self.height)
+
+	@property
+	def topLeft(self):
+		return Point(self.left,self.top)
+
+	@property
+	def topRight(self):
+		return Point(self.right,self.top)
+
+	@property
+	def bottomLeft(self):
+		return Point(self.left,self.bottom)
+
+	@property
+	def bottomRight(self):
+		return Point(self.right,self.bottom)
+
+	@property
+	def center(self):
+		return Point(int(round(self.left+self.width/2.0)), int(round(self.top+self.height/2.0)))
+
+	def __contains__(self,other):
+		"""Returns whether other is a part of this rectangle."""
+		if isinstance(other,POINT_CLASSES):
+			return self.left <= other.x < self.right and  self.top <= other.y < self.bottom
+		if not isinstance(other,RECT_CLASSES):
+			try:
+				other=toRectLTRB(other)
+			except:
+				return False
+		return self.isSuperset(other) and self!=other
+
+	def isSubset(self,other):
+		"""Returns whether this rectangle is a subset of other (i.e. whether all points in this rectangle are contained by other)."""
+		if not isinstance(other,RECT_CLASSES):
+			try:
+				other=toRectLTRB(other)
+			except ValueError:
+				return False
+		return self.left >= other.left and self.top >= other.top and self.right <= other.right and self.bottom <= other.bottom
+
+	def isSuperset(self,other):
+		"""Returns whether this rectangle is a superset of other (i.e. whether all points of other are contained by this rectangle)."""
+		if not isinstance(other,RECT_CLASSES):
+			try:
+				other=toRectLTRB(other)
+			except ValueError:
+				return False
+		return other.left >= self.left and other.top >= self.top and other.right <= self.right and other.bottom <= self.bottom
+
+	def __eq__(self,other):
+		if not isinstance(other,RECT_CLASSES):
+			try:
+				other=toRectLTRB(other)
+			except ValueError:
+				return False
+		return other.left == self.left and other.top == self.top and other.right == self.right and other.bottom == self.bottom

```

Comment:
Maybe this has already been considered, but it has occured to me that if we are mixing rectangle like classes that come from a source that uses the top left as the origin with those that use the bottom left then given a screen space 10 heigh (indexes 0 to 9):
top of the former is index 3, the top of the latter is index 6 then top should be the same.

Reply:
I belief we aren't supporting rectangles that are using the bottom left as their origin.

Reply:
Having said that, wx.Rect seems to be incompatible, so I"m removing that one from the supported classes.

```
r=wx.Rect(10,10,11,11)
>>> r.top
10
>>> r.left
10
>>> r.width
11
>>> r.height
11
>>> r.right
20
>>> r.bottom
20
```

Reply:
I don't really understand what is going on in that `wx.Rect` example? My guess is that `right` and `bottom` are **inclusive** rather than our `Rect` class which are exclusive?

As I mentioned in my other comment, I think we should add a way to mix coordinate spaces, and force those constructing `Rect` classes to consider what coordinate space they are in. Even if all the types they are using are in the same space, if you don't consider it you may be visualising things wrong and not get the results that you expect (as I did, even though, with hindsight, I know I should have considered it)